### PR TITLE
Merging master with data-publisher-integration

### PR DESCRIPTION
**Pull request by Thanu**
This P/R is to merge master branch changes into data-publisher-integration.

### CODE REVIEW

File: components/org.apache.stratos.autoscaler/src/main/java/org/apache/stratos/autoscaler/applications/parser/DefaultApplicationParser.java [hunk 1]
```diff
@@ -455,7 +455,7 @@ private Application buildCompositeAppStructure(ApplicationContext applicationCon
                     subscribableInfoContext.getDeploymentPolicy(), isLB,
                     tenantRange, subscribableInfoContext.getDependencyAliases(),
                     subscribableInfoContext.getProperties(), arrDependencyClusterIDs, arrExportMetadata,
-                    arrImportMetadata);
+                    arrImportMetadata,subscribableInfoContext.getLvsVirtualIP());
```
Comment (lahirus) [inline components/org.apache.stratos.autoscaler/src/main/java/org/apache/stratos/autoscaler/applications/parser/DefaultApplicationParser.java]: minor formatting issues.  Space missing after comma.

File: components/org.apache.stratos.autoscaler/src/main/java/org/apache/stratos/autoscaler/applications/parser/DefaultApplicationParser.java [hunk 2]
```diff
@@ -942,12 +942,12 @@ private ApplicationClusterContext createApplicationClusterContext(String appId,
                                                                       String alias, String clusterId, String hostname,
                                                                       String deploymentPolicy, boolean isLB, String tenantRange,
                                                                       String[] dependencyAliases, Properties properties, String[] dependencyClustorIDs,
-                                                                      String[] exportMetadata, String[] importMetadata)
+                                                                      String[] exportMetadata, String[] importMetadata,String lvsVirtualIP)
```
Comment (lahirus) [inline components/org.apache.stratos.autoscaler/src/main/java/org/apache/stratos/autoscaler/applications/parser/DefaultApplicationParser.java]: same issue mentioned above

File: components/org.apache.stratos.autoscaler/src/main/java/org/apache/stratos/autoscaler/applications/parser/DefaultApplicationParser.java [hunk 3]
```diff
@@ -942,12 +942,12 @@ private ApplicationClusterContext createApplicationClusterContext(String appId,
...
             throws ApplicationDefinitionException {
 
         // Create text payload
         PayloadData payloadData = ApplicationUtils.createPayload(appId, groupName, cartridge, subscriptionKey, tenantId, clusterId,
-                hostname, repoUrl, alias, null, dependencyAliases, properties, oauthToken, dependencyClustorIDs, exportMetadata, importMetadata);
+                hostname, repoUrl, alias, null, dependencyAliases, properties, oauthToken, dependencyClustorIDs, exportMetadata, importMetadata,lvsVirtualIP);
```
Comment (lahirus) [inline components/org.apache.stratos.autoscaler/src/main/java/org/apache/stratos/autoscaler/applications/parser/DefaultApplicationParser.java]: Formatting seems wrong. Pls check.

File: components/org.apache.stratos.autoscaler/src/main/java/org/apache/stratos/autoscaler/context/InstanceContext.java [hunk 4]
```diff
@@ -29,34 +29,19 @@
  * This will hold the instances related info.
  */
 public abstract class InstanceContext {
+    //instance id
     protected String id;
-
+    //parent instance id
```
Comment (lahirus) [inline components/org.apache.stratos.autoscaler/src/main/java/org/apache/stratos/autoscaler/context/InstanceContext.java]: this comment is not useful. variable name is same.

File: components/org.apache.stratos.autoscaler/src/main/java/org/apache/stratos/autoscaler/context/cluster/ClusterInstanceContext.java [hunk 5]
```diff
@@ -40,22 +40,21 @@
  * It holds the runtime data of a VM cluster
  */
 public class ClusterInstanceContext extends InstanceContext {
-
     private static final Log log = LogFactory.getLog(ClusterInstanceContext.class);
-
+    //partition algorithm
```
Comment (lahirus) [inline components/org.apache.stratos.autoscaler/src/main/java/org/apache/stratos/autoscaler/context/cluster/ClusterInstanceContext.java]: comment is not useful.

File: components/org.apache.stratos.autoscaler/src/main/java/org/apache/stratos/autoscaler/context/partition/ClusterLevelPartitionContext.java [hunk 6]
```diff
@@ -58,6 +58,8 @@
 
     // 15 mints as the default
     private long pendingMemberExpiryTime = 900000;
+    //
```
Comment (lahirus) [inline components/org.apache.stratos.autoscaler/src/main/java/org/apache/stratos/autoscaler/context/partition/ClusterLevelPartitionContext.java]: remove the commented empty line.

File: components/org.apache.stratos.autoscaler/src/main/java/org/apache/stratos/autoscaler/context/partition/ClusterLevelPartitionContext.java [hunk 7]
```diff
@@ -58,6 +58,8 @@
 
     // 15 mints as the default
```
Comment (lahirus) [inline components/org.apache.stratos.autoscaler/src/main/java/org/apache/stratos/autoscaler/context/partition/ClusterLevelPartitionContext.java]: comments spelling can be improved

**Comment by Thanu** [comment]
I haven't check any formatting issue. This is just to merge master branch changes to data-publisher-integration. IMO it's better to do this in master branch.

File: components/org.apache.stratos.autoscaler/src/main/java/org/apache/stratos/autoscaler/monitor/component/GroupMonitor.java [hunk 8]
```diff
@@ -793,13 +780,13 @@ private void addPartitionContext(Instance parentInstanceContext,
      * Creates the group instance and adds the required context objects
      *
      * @param group                              the group
-     * @param parentInstanceContext              the parent instance context
+     * @param parentInstance              the parent instance context
```
Comment (lahirus) [inline components/org.apache.stratos.autoscaler/src/main/java/org/apache/stratos/autoscaler/monitor/component/GroupMonitor.java]: minor formatting issue in parentInstance param

File: components/org.apache.stratos.common/src/main/java/org/apache/stratos/common/constants/StratosConstants.java [hunk 9]
```diff
@@ -91,7 +91,8 @@
 
     // metering constants
     public static final String THROTTLING_ALL_ACTION = "all_actions";
-    public static final String THROTTLING_IN_DATA_ACTION = "in_data_action"; //this covers registry capacity + registry bandwidth
+    public static final String THROTTLING_IN_DATA_ACTION =
```
Comment (sinthuja) [inline components/org.apache.stratos.common/src/main/java/org/apache/stratos/common/constants/StratosConstants.java]: Formatting issue, it's not needed to be in the second line.

File: products/stratos/modules/distribution/src/main/conf/drools/scaling.drl [hunk 10]
```diff
@@ -162,6 +162,10 @@ dialect "mvel"
                     boolean partitionsAvailable = true;
                     int count = 0;
 
+                    String autoscalingReason = (numberOfRequiredInstances == numberOfInstancesReuquiredBasedOnRif)?"Scaling up due to RIF, [Predicted Value] "+rifPredictedValue+" [Threshold] "+rifThreshold:(numberOfRequiredInstances== numberOfInstancesReuquiredBasedOnMemoryConsumption)?"Scaling up due to MC, [Predicted Value] "+mcPredictedValue+" [Threshold] "+mcThreshold:"Scaling up due to LA, [Predicted Value] "+laPredictedValue+" [Threshold] "+laThreshold;
+                    autoscalingReason += " [Number of required instances] "+numberOfRequiredInstances+" [Cluster Max Members] "+clusterMaxMembers+" [Additional instances to be created] " + additionalInstances;
+
```
Comment (sinthuja) [inline products/stratos/modules/distribution/src/main/conf/drools/scaling.drl]: Remove unnecessary new lines.

File: components/org.apache.stratos.common/src/main/java/org/apache/stratos/common/constants/StratosConstants.java (hunk 9 above)
Comment (Thanu) [inline components/org.apache.stratos.common/src/main/java/org/apache/stratos/common/constants/StratosConstants.java]: Thanks for pointing this. This is now fixed in master.

**Pull request merged**